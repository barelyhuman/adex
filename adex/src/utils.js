export { default as ms } from 'ms'
export { default as Youch } from 'youch'
