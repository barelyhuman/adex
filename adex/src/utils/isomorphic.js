export { parse as pathToRegex } from 'regexparam'
export { renderToStringAsync } from 'preact-render-to-string'
