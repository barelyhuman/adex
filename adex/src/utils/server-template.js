/**
 * Returns the standard server module template for adex
 *
 * @param {string} adapter - The adapter to use for the server
 * @returns {string} - The server template code
 */
export function getServerTemplate(adapter) {
  return `import { createServer } from '${adapter}'
import { dirname, join } from 'node:path'
import { fileURLToPath } from 'node:url'
import { existsSync, readFileSync } from 'node:fs'
import { env } from 'adex/env'

import 'virtual:adex:font.css'
import 'virtual:adex:global.css'

const __dirname = dirname(fileURLToPath(import.meta.url))

const PORT = parseInt(env.get('PORT', '3000'), 10)
const HOST = env.get('HOST', 'localhost')

const paths = {
  assets: join(__dirname, './assets'),
  islands: join(__dirname, './islands'),
  client: join(__dirname, '../client'),
}

function getServerManifest() {
  const manifestPath = join(__dirname, 'manifest.json')
  if (existsSync(manifestPath)) {
    const manifestFile = readFileSync(manifestPath, 'utf8')
    return parseManifest(manifestFile)
  }
  return {}
}

function getClientManifest() {
  const manifestPath = join(__dirname, '../client/manifest.json')
  if (existsSync(manifestPath)) {
    const manifestFile = readFileSync(manifestPath, 'utf8')
    return parseManifest(manifestFile)
  }
  return {}
}

function parseManifest(manifestString) {
  try {
    const manifestJSON = JSON.parse(manifestString)
    return manifestJSON
  } catch (err) {
    return {}
  }
}

const server = createServer({
  port: PORT,
  host: HOST,
  adex:{
    manifests:{server:getServerManifest(),client:getClientManifest()},
    paths,
  }
})

if ('run' in server) {
  server.run()
}

export default server.fetch
`
}
