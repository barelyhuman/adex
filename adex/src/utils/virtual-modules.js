/**
 * Creates a virtual module that can be imported in Vite
 *
 * @param {string} id - The module ID
 * @param {string} content - The module content
 * @returns {import("vite").Plugin}
 */
export function createVirtualModule(id, content) {
  return {
    name: `adex-virtual-${id}`,
    enforce: 'pre',
    resolveId(requestId) {
      if (requestId === id || requestId === '/' + id) {
        return `\0${id}`
      }
    },
    load(requestId) {
      if (requestId === `\0${id}`) {
        return content
      }
    },
  }
}

/**
 * Creates a virtual module that can fall back to a user-provided file
 *
 * @param {string} id - The virtual module ID
 * @param {string} content - The default content if user file doesn't exist
 * @param {string} userPath - The path to the user file
 * @returns {import("vite").Plugin}
 */
export function createUserDefaultVirtualModule(id, content, userPath) {
  return {
    name: `adex-virtual-user-default-${id}`,
    enforce: 'pre',
    async resolveId(requestId) {
      if (
        requestId === id ||
        requestId === '/' + id ||
        requestId === userPath
      ) {
        const userPathResolved = await this.resolve(userPath)
        return userPathResolved ?? `\0${id}`
      }
    },
    load(requestId) {
      if (requestId === `\0${id}`) {
        return content
      }
    },
  }
}
