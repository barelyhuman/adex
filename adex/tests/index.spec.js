import { beforeAll, describe, expect, it } from 'vitest'

import { devServerURL, launchDemoDevServer } from './utils.js'

describe('ssr minimal', () => {
  beforeAll(async () => {
    await launchDemoDevServer('tests/fixtures/minimal')
  })

  it('basic response', async () => {
    const response = await fetch(devServerURL).then(d => d.text())
    expect(response).toMatchSnapshot()

    const response2 = await fetch(new URL('/about', devServerURL)).then(d =>
      d.text()
    )
    expect(response2).toMatchSnapshot()
  })
})
