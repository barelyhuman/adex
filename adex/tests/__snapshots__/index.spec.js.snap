// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ssr minimal > basic response 1`] = `
"
    <!doctype html>
    <html lang="">
      <head>
        <script type="module" src="/@vite/client"></script>

        <meta charset="UTF-8" />
        
    <title></title>

    
    
  
      </head>
      <body>
        <div id="app"><h1>Hello World</h1></div>
      <script type='module' src="/virtual:adex:client"></script></body>
    </html>
  "
`;
